cmd/app/main.go

```diff
@@ -60,11 +66,47 @@ func main() {
 		log.Fatal(err)
 	}
 
-	server := virtual.NewServer(reg, environment)
+	var server Server = virtual.NewServer(reg, environment)
 
 	log.Printf("listening on port: %d\n", *port)
 
-	if err := server.Start(*port); err != nil {
+	go func(server Server) {
+		sig := waitForSignal()
+		log.Printf("received signal: %s", sig.String())
+		shutdown(server, *shutdownTimeout)
+	}(server)
+
+	if err := server.Start(*port); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		log.Printf("received error: %s", err.Error())
+		shutdown(server, *shutdownTimeout)
 		log.Fatal(err)
 	}
 }
+
+type Server interface {
```

Comment:
Why is this interface needed and why is it public? It cant be imported since its a main package

Reply:
I made it private., The reason why it is needed is that `virtual.server` is private